### ISSUE TITLE
OpenMS binaries are too large

### ISSUE THREAD

**Issue by hroest** [OpenMS/OpenMS#5619]
I realized that in the latest release (2.7) we produce 532 MB of binaries in the `bin` folder when statically linking. The reason for this is that the majority of our binaries statically link the xercesc library and include about 3.8 MB of code, in our latest release this affected 139 binaries accounting for around 500 MB of the size of the folder. This also affects the latest release OpenMS-2.7.0-Debian-Linux-x86_64.deb which *should* in principle have been dynamically linked but wasn't, looking at the file with `ar x` and extracting `data.tar.gz` I find 546 MB in the `bin` folder. 

I see two solutions to this
- dynamically link our binaries (when I do this, I get around 20 MB for the binaries)
- remove the need for linking `xercesc` by keeping it inside the `OpenMS::Internal` namespace

I dont know how feasible the first solution is, especially on Windows but we are clearly duplicating a lot of code by linking it into 140 different binaries.

**Comment by hroest** [comment]
Yes, have a look https://github.com/OpenMS/OpenMS/blob/develop/src/openms/include/OpenMS/FORMAT/HANDLERS/XMLHandler.h -- in its current implementation there is no way to separate XMLHandler from `xercesc` since it inherits from `xercesc::DefaultHandler` and that is how the parser interface works, you have to inherit from xerces. We therefore have to hide the implementation by either hiding the XMLFile itself through `FileHandler` or by hiding the implementation in a "Handler" class.

**Comment by hroest** [comment]
> Totally agree on your second point. So the catches are xerces objects and full declarations in our public API right?

@jpfeuffer yes this is what gets included

https://github.com/OpenMS/OpenMS/blob/8fb42eca245b70bc79818a4f37e6243f67225c1a/src/openms/include/OpenMS/FORMAT/HANDLERS/XMLHandler.h#L45-L48

so a whole bunch of stuff in our public API and requiring downstream tools to link against xerces to make sure these symbols are defined. Also we inherit from `xercesc::DefaultHandler` so there is no way to hide this in the .cpp file

**Comment by jpfeuffer** [comment]
Ok so switch from inheritance to composition and expose only functions that do not have xerces types?

And put some of the templated helper functions into cpp by specializing on the few datatypes we need?

**Comment by jpfeuffer** [comment]
But I guess we *have to* inherit from the Xerces handler class, right? I found no other pattern to use Xerces. This means we have to hide our Handler completely.

**Comment by jpfeuffer** [comment]
Would it be enough to remove OPENMS_DLLAPI from our XMLHandler class since the default visibility preset is "hidden"?

**Comment by hroest** [comment]
@jpfeuffer yes, we have to inherit from the xerces handler in order to get the callback right. I dont think tinkering with OPENMS_DLLAPI would be the right approach since i) we still need to expose classes such as `ProtXMLFile` to the public API, so we cannot remove them and ii) I am not sure if this would even help as long as we include these headers and use the classes. Imagine a TOPP tool that uses `ProtXMLFile`.  `ProtXMLFile` inherits from `Internal::XMLHandler` which inherits from `xercesc` and therefore any downstream user of ProtXMLFile needs to know about all these xerces classes otherwise the ProtXMLFile is not completely defined.
